### ISSUE TITLE
Channel failure due to unknown issue with eclair

### ISSUE THREAD

**Issue by viaj3ro** [ACINQ/eclair#2070]
I had a channel failure recently which according to Anton from BLW seems to be caused by an unknown issue with eclair. Diskussion here: https://t.me/eclair_lightning

```
2021-11-09 07:18:23,908 INFO  f.a.e.b.b.ZmqWatcher - triggering WatchFundingSpent(Actor[akka://eclair-node/user/$h/switchboard/peer-03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315/$a#-996100026],62c8129b56bc0622ffcd26cb058e49e8691003fffabf35977f1df0c44ec905b4,1,Set(54d6ea4e322496d7c222583af1012be2c656f95395683fd2fcee9a26f133026a, c3edafc8d79833437890ec68b3c2a72c367949c15ea7f455709e9d186797e758))
2021-11-09 07:18:23,947 WARN  f.a.e.channel.Channel n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - they published their current commit in txid=0d2eb24a2e60f1a6143db02aef7ec517351363be6f082585395a3872cf9f5b0b
2021-11-09 07:18:23,948 INFO  f.a.e.db.DbEventHandler n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - paying mining fee=0 sat for txid=0d2eb24a2e60f1a6143db02aef7ec517351363be6f082585395a3872cf9f5b0b desc=remote-commit
2021-11-09 07:18:23,965 INFO  f.a.e.channel.Channel n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - tx generation success: desc=claim-htlc-success txid=e7d779799a41f4ad56874957ce94723b3c0c17d1efccaa786da8eaf07fd3e761 amount=12858 sat tx=02000000010b5b9fcf72385a398525086fbe63133517c57eef2ab03d14a6f1602e4ab22e0d0200000000ffffffff013a320000000000001600141d1fd469ec50c8013c853b5eb91bd26761da9c7300000000
2021-11-09 07:18:23,967 INFO  f.a.e.channel.Channel n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - tx generation success: desc=claim-htlc-success txid=df3d5460fb28f6c9322429df66abc0480b8b3f5fdf2075b935075da0e4816804 amount=12858 sat tx=02000000010b5b9fcf72385a398525086fbe63133517c57eef2ab03d14a6f1602e4ab22e0d0100000000ffffffff013a320000000000001600141d1fd469ec50c8013c853b5eb91bd26761da9c7300000000
2021-11-09 07:18:23,967 INFO  f.a.e.channel.Channel n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - tx generation skipped: desc=local-anchor reason: output not found (probably trimmed)
2021-11-09 07:18:23,968 INFO  f.a.e.channel.Channel n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - tx generation skipped: desc=remote-anchor reason: output not found (probably trimmed)
2021-11-09 07:18:24,000 INFO  f.a.e.b.b.ZmqWatcher - 0d2eb24a2e60f1a6143db02aef7ec517351363be6f082585395a3872cf9f5b0b:1 has already been spent, looking for the spending tx in the mempool
``` 

Happy to provide more logs via telegram

**Comment by t-bast** [comment]
Interesting, since you are indeed broadcasting an htlc-success transaction it means you had the preimage, but for some reason the HTLC was not settled. I am indeed interested to get more logs.

Can you search your log to figure out which htlc you're claiming on-chain, then locate the time when you received the corresponding preimage for this htlc (it must have been fulfilled downstream) and send me all the logs around that time. Please send the logs to me via email or upload them somewhere and provide me with the link.

**Comment by viaj3ro** [comment]
what's the best way to go about this? what search strings and which dates?

**Comment by t-bast** [comment]
Thanks to @viaj3ro's logs, I've been able to troubleshoot this.
The TL;DR is that there is nothing too weird here and nobody lost money.

The issue is that an HTLC timed out downstream, but the HTLC-timeout transaction took too long to confirm.
We cannot safely propagate the failure upstream until the downstream HTLC-timeout transaction is confirmed.
Unfortunately, the upstream HTLC timed out before the downstream HTLC-timeout transaction was confirmed, so the upstream channel had to close too.

With anchor outputs, we would have been able to make the HTLC-timeout transaction confirm faster and avoid the upstream close.

Here are all the details of how I investigated it, so that in the future, node operator can learn how to investigate themselves.

---

We start by filtering only the logs for this channel:

```sh
grep -i b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 eclair.2021-11-09.log
```

This finds the following log line:

```log
2021-11-09 07:15:29,698 ERROR f.a.e.channel.Channel n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - peer sent error: ascii='one or more htlcs timed out downstream: ids=786' bin=6f6e65206f72206d6f72652068746c63732074696d6564206f757420646f776e73747265616d3a206964733d373836
```

So we now know we're looking for HTLC #786, so we search for when it was received:

```sh
grep -i "IN msg=UpdateAddHtlc(b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863,786" eclair.2021-11-09.log
```

This doesn't return any result, indicating that the HTLC was received before the beginning of these logs.
We run this command on previous days' logs.
We then find this HTLC in the `eclair.2021-11-05.log` file.
Finding this log line reveals that the corresponding payment hash is `25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc`.

We now filter logs for that payment hash.
Notice that we only use the first 7 characters, because the payment hash is usually shortened to avoid long log lines.

```sh
grep -i 25335ab eclair.2021-11-05.log
```

We notice that this payment hash was attempted multiple times, relayed, and then failed:

```log
2021-11-05 19:21:19,311 INFO  f.a.eclair.Diagnostics PAY n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - IN msg=UpdateAddHtlc(b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863,778,15005722 msat,25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc,CltvExpiry(708872),OnionRoutingPacket(0,ByteVector(33 bytes, 0x02c40c556b23e703d81f0f56e54383151542408ca9c2b2e9cb1ac5dafe2d5f7e61),ByteVector(1300 bytes, #-360722058),0d5aeaf4985a06e3431fbf3e140b15ef9c2f729b1deccf572d911c0bb36d7505),TlvStream(List(),List()))
2021-11-05 19:21:19,394 INFO  f.a.e.p.r.ChannelRelay PAY h:25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc p:d959e88b - relaying htlc #778 from channelId=b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 to requestedShortChannelId=690025x3035x0 nextNode=030c3f19d742ca294a55c00376b3b355c3c90d61c6b6b39554dbc7ac19b141c14f
2021-11-05 19:21:19,396 INFO  f.a.e.i.Peer PAY n:030c3f19d742ca294a55c00376b3b355c3c90d61c6b6b39554dbc7ac19b141c14f c:e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a - OUT msg=UpdateAddHtlc(e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a,3687,15005690 msat,25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc,CltvExpiry(708822),OnionRoutingPacket(0,ByteVector(33 bytes, 0x032b8f0712c096137a428aea2dd985b16bcf9b90ccbf16ec3272c97291979ce739),ByteVector(1300 bytes, #-889457239),22d59d4d344ab10610fe3f22b9491b6360e84efb8b9586eca4e2b5fd10595b55),TlvStream(List(),List()))
2021-11-05 19:22:15,156 INFO  f.a.e.p.r.ChannelRelay PAY h:25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc p:d959e88b - relaying fail to upstream
```

However, for HTLC #786, it was relayed out but not failed (at least not on that day):

```log
2021-11-05 19:34:25,933 INFO  f.a.e.i.Peer PAY n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - IN msg=UpdateAddHtlc(b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863,786,15005722 msat,25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc,CltvExpiry(708872),OnionRoutingPacket(0,ByteVector(33 bytes, 0x020b417b259077c441de612046f6339f55942310b22d49666c931fb1c20b6e8632),ByteVector(1300 bytes, #-524020781),57ce0e7a1316459d2586e9929d5ccceeb3bfa99c771976468abcf8e7d6bb9a7d),TlvStream(List(),List()))
2021-11-05 19:34:26,034 INFO  f.a.e.p.r.ChannelRelay PAY h:25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc p:34d924d2 - relaying htlc #786 from channelId=b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 to requestedShortChannelId=690025x3035x0 nextNode=030c3f19d742ca294a55c00376b3b355c3c90d61c6b6b39554dbc7ac19b141c14f
2021-11-05 19:34:26,036 INFO  f.a.e.i.Peer PAY n:030c3f19d742ca294a55c00376b3b355c3c90d61c6b6b39554dbc7ac19b141c14f c:e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a - OUT msg=UpdateAddHtlc(e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a,3695,15005690 msat,25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc,CltvExpiry(708822),OnionRoutingPacket(0,ByteVector(33 bytes, 0x027535be512f303cb17a14423420b701e68c9342e5289e310823e5cfb7b4905a15),ByteVector(1300 bytes, #1278800489),f46af1ee606e66c184881a01f482020dc5232f29c60b3676d3f3d02eb6e10a84),TlvStream(List(),List()))
```

The corresponding outgoing HTLC is #3695 in channel `e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a`, let's see if that payment was fulfilled:

```sh
grep -i "UpdateFulfillHtlc(e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a,3695" eclair.2021-11-05.log
grep -i "UpdateFulfillHtlc(e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a,3695" eclair.2021-11-06.log
grep -i "UpdateFulfillHtlc(e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a,3695" eclair.2021-11-07.log
grep -i "UpdateFulfillHtlc(e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a,3695" eclair.2021-11-08.log
grep -i "UpdateFulfillHtlc(e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a,3695" eclair.2021-11-09.log
```

None of these return any result, indicating that this outgoing HTLC was not fulfilled.
Let's have a look at the outgoing channel `e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a`.
We search for errors in this channel:

```sh
grep ERROR eclair.2021-11-0*.log
```

We notice that the channel is closing, because an HTLC timed out downstream:

```log
2021-11-07 21:54:02,976 ERROR f.a.e.channel.Channel n:030c3f19d742ca294a55c00376b3b355c3c90d61c6b6b39554dbc7ac19b141c14f c:e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a - one or more htlcs timed out downstream: ids=3728 while processing msg=CurrentBlockCount in state=OFFLINE
```

We now search for this specific HTLC to see whether it has been claimed on-chain:

```sh
grep -i "htlc #3695" eclair.2021-11-0*.log
```

Which shows that it has been correctly timed out on-chain:

```log
2021-11-09 11:17:54,070 INFO  f.a.e.channel.Channel n:030c3f19d742ca294a55c00376b3b355c3c90d61c6b6b39554dbc7ac19b141c14f c:e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a - htlc-timeout tx for htlc #3695 paymentHash=25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc expiry=708822 has been confirmed (tx=020000000001013b7677a41a51df503456c10cae99ee6bc806d385089aff65738b717dec7104b10100000000000000000151390000000000002200209fa9bd8687f53bf5e33bca61072defe70dafd22e4fb8b905d1ce7b27918c30080500473044022017cb09746ce3d9fe8ffe9457668d55aee6221035137833e0cddc3c426a0f26810220283b57c9edae897570d8528e825cfadc61e6bd552d849b9a69e072683eb139ba01473044022069cf355771fee82e6ba37541cc8c6bf73dcebfef6d57426ee52cc9055d42bd880220558aa43230aced484661f5a1efa3807b4d1f52a3f299dc7b3b1a22ef6ab42faa01008576a9146e554542cb2a605f5f4731e8ee8339f46dd393e78763ac672103958485257303fb342cf9a1967768c99b9b16c62e05f20f6ad472ae5c2b5b221c7c820120876475527c2102ea873720c043308587d972b616c78af6d31c5d08b01ca35cbb962ea5dd9d4ef752ae67a914515b9994a44491a7c756c4f92b31fa42e58c085688ac6868d6d00a00)
2021-11-09 11:17:54,070 INFO  f.a.e.channel.Channel n:030c3f19d742ca294a55c00376b3b355c3c90d61c6b6b39554dbc7ac19b141c14f c:e0c5816d9c09ce6786e5f785d3044f9cce99b0b20d8df4e71cd21dc2cb3f008a - failing htlc #3695 paymentHash=25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc origin=ChannelRelayedHot(Actor[akka://eclair-node/user/$g/relayer/channel-relayer/34d924d2-0a44-4436-ae6e-8cc7b5539692/$$a-adapter#1785278545],UpdateAddHtlc(b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863,786,15005722 msat,25335ab6932fdd40feb646cff4aadabfda88180542f3d683b2efcc7369e4dbdc,CltvExpiry(708872),OnionRoutingPacket(0,ByteVector(33 bytes, 0x020b417b259077c441de612046f6339f55942310b22d49666c931fb1c20b6e8632),ByteVector(1300 bytes, #-524020781),57ce0e7a1316459d2586e9929d5ccceeb3bfa99c771976468abcf8e7d6bb9a7d),TlvStream(List(),List())),15005690 msat): htlc timed out
```

We notice that while the channels closed at `2021-11-07 21:54:02,976`, the HTLC-timeout transaction only confirmed at `2021-11-09 11:17:54,070`.
At that point, we would have propagated an `UpdateFailHtlc` to the upstream channel (`b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863`), but unfortunately it's too late: this channel was closed at `2021-11-09 07:15:29,698`.

The commitment feerate was too low and we had no way of doing CPFP to speed up confirmation, because the channel was not using anchor outputs.
With anchor outputs, we would have RBF-ed the HTLC-timeout transaction to get it to confirm faster, and would have been able to avoid the upstream channel closure.

There is one thing that is still worth looking into. The HTLC was timed out downstream, but the following log lines make it look like we had the preimage:

```log
eclair.2021-11-09.log:2021-11-09 07:18:23,965 INFO  f.a.e.channel.Channel n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - tx generation success: desc=claim-htlc-success txid=e7d779799a41f4ad56874957ce94723b3c0c17d1efccaa786da8eaf07fd3e761 amount=12858 sat tx=02000000010b5b9fcf72385a398525086fbe63133517c57eef2ab03d14a6f1602e4ab22e0d0200000000ffffffff013a320000000000001600141d1fd469ec50c8013c853b5eb91bd26761da9c7300000000
eclair.2021-11-09.log:2021-11-09 07:18:23,967 INFO  f.a.e.channel.Channel n:03ee58475055820fbfa52e356a8920f62f8316129c39369dbdde3e5d0198a9e315 c:b405c94ec4f01d7f9735bffaff031069e8498e05cb26cdff2206bc569b12c863 - tx generation success: desc=claim-htlc-success txid=df3d5460fb28f6c9322429df66abc0480b8b3f5fdf2075b935075da0e4816804 amount=12858 sat tx=02000000010b5b9fcf72385a398525086fbe63133517c57eef2ab03d14a6f1602e4ab22e0d0100000000ffffffff013a320000000000001600141d1fd469ec50c8013c853b5eb91bd26761da9c7300000000
```

However, it turns out that this is a red herring: we always generate the claim-htlc-success transaction (https://github.com/ACINQ/eclair/blob/6cc37cbd4f8192162e916ddabe3a612fafed06a2/eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala#L752) and only afterwards we check whether we have the preimage to broadcast this tx (https://github.com/ACINQ/eclair/blob/6cc37cbd4f8192162e916ddabe3a612fafed06a2/eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala#L754).

The logs don't say that we published these `claim-htlc-success` transaction, which means we didn't have the preimage.
Eclair should avoid generating this confusing log line though, that will be fixed.
